unified_planning/io/anml_writer.py

```diff
@@ -0,0 +1,329 @@
+            return str(dec)
+
+    def walk_int_constant(self, expression, args):
+        assert len(args) == 0
+        return str(expression.constant_value())
+
+    def walk_plus(self, expression, args):
+        assert len(args) > 1
+        return f"({' + '.join(args)})"
+
+    def walk_minus(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} - {args[1]})'
+
+    def walk_times(self, expression, args):
+        assert len(args) > 1
+        return f"({' * '.join(args)})"
+
+    def walk_div(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} / {args[1]})'
+
+    def walk_le(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} <= {args[1]})'
+
+    def walk_lt(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} < {args[1]})'
+
+    def walk_equals(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} == {args[1]})'
+
+
+class ANMLWriter:
+    '''This class can be used to write a Problem in ANML.'''
+
+    def __init__(self, problem: 'up.model.Problem'):
+        self.problem = problem
+
+    def _write_problem(self, out: IO[str]):
+        types_mapping: Dict['up.model.Type', str] = {}
+        # Init types_mapping.
+        types_mapping[self.problem.env.type_manager.IntType()] = 'integer'
+        types_mapping[self.problem.env.type_manager.RealType()] = 'rational'
+        for t in self.problem.user_types:
+            ut = cast(_UserType, t)
+            if _is_valid_anml_name(ut.name): # No renaming needed
+                types_mapping[t] = ut.name
+
+        for t in self.problem.all_types:
+            #NOTE If ANML does not accept a re-defintion of 'integer' or 'rational' or maybe also 'boolean', they must be skipped
+            anml_type_name = _get_mangled_type_name(t, types_mapping)
+            out.write(f'type {anml_type_name}')
+            if t.is_bool_type() or (t.is_user_type() and cast(_UserType, t).father is None):
+                out.write(';\n')
+            elif t.is_user_type(): # and cast(_UserType, t).father is not None:
+                # For construction in the Problem, the father of a UserType is always added before the UserType itself.
+                father = cast(_UserType, t).father
+                assert father is not None
+                assert types_mapping[father] is not None
+                out.write(f' < {types_mapping[father]};\n')
+            elif (t.is_int_type() or t.is_real_type()) and cast(_IntType, t).lower_bound is None and cast(_IntType, t).upper_bound is None: #If first NOTE is True, this if is impossible
+                out.write(';\n')
+            elif t.is_int_type() and cast(_IntType, t).upper_bound is None: # and t.lower is not None
+                out.write(f' < integer [{str(cast(_IntType, t).lower_bound)}, infinity);\n')
```

Comment:
What is this syntax? There is no need for sub-typing here just "integer[123, infinity]" represents the type.